### DISCUSSION TITLE
How to dispatch different for single device?

### DISCUSSION THREAD

**Discussion by jyu99** [fwupd/fwupd#2794] [General]
We have wireless product , just like  a set of dongle , mouse , and keyboard. Actually, fwupdmgr  only dectected dongle device.
If I want to update mouse firmware ,any solution about this problem? e.g. if I upload the mouse firmware , the fwupdmgr should show the firmware release? use the metadata for `Composite Hardware` in document or other way to reach this function?

**Comment by hughsie** [comment]
> If I want to update mouse firmware ,any solution about this problem

Sure, we do this already for Logitech unifying devices.

> fwupdmgr should show the firmware release

You'd need to add it to the existing plugin (or write a new plugin) which can enumerate the devices and also send them the new firmware. I'm assuming you have some code already that can do this?

**Reply by jyu99** [reply]
Thank your reply

**Reply by jyu99** [reply]
I know we can't get the release information in normal way ,but I still wanna ask that current fwupd exist a tricky way to get the release information? 
Just like version number?

**Reply by hughsie** [reply]
I'm not sure how to answer that; there are hundreds of different ways to get a release version. If I understand correctly, Collabera are planning to work on some of the btle stuff in fwupd soon. I think that will be useful for you.

**Comment by jyu99** [comment]
I  have dongle. It has two devices ,wireless mouse and wireless  keyboard. In practical only dongle connected to PC. I create two child  FuDevice as below

<pre>├─256 MOUSE:
│ │   Device ID:          d56fc8e39a01425d5a15bd7ffc8664cfd40d3d3c
│ │   Current version:    1.0.0
│ │   Vendor:             USB:0x093A|HIDRAW:0x093A
│ │   GUIDs:              e84daf50-dfb6-5800-807b-1fb4aa54a4a8 ← HIDRAW\VEN_093A&amp;DEV_2801&amp;REV_00
│ │                       7bd6065d-14d7-54a5-9487-428d66c18936 ← HIDRAW\VEN_093A&amp;DEV_2801
│ │                       46eb63e2-a2d8-5ac4-b77e-f86886b2bb8f ← HIDRAW\VID_093A&amp;PID_2801&amp;DEV_256 MOUSE&amp;NUM129
│ │   Device Flags:       • Updatable
│ │                       • Supported on remote server
│ │ 
│ ├─CHILD MOUSE:
│ │     Device ID:        87835653f7f79d15420fc69988bdadcd623b1314
│ │     Current version:  1.5.6
│ │     Vendor:           USB:0x093A
│ │     GUID:             dbb64836-12f6-519e-a439-904ef0bff587 ← HIDRAW\VID_093A&amp;PID_2801&amp;IDX2
│ │     Device Flags:     • Updatable
│ │                       • Supported on remote server
│ │   
│ └─CHILD KB:failed to open device for prepare: permission denied

│       Device ID:        3b42553c4e3241e8f3f8fbc19a69fa2f95708a9d
│       Current version:  1.5.6
│       Vendor:           USB:0x093A
│       GUID:             569fa7b4-776e-535d-b369-5cab2e4c891d ← HIDRAW\VID_093A&amp;PID_2801&amp;IDX1
│       Device Flags:     • Updatable
│                         • Supported on remote server
</pre>

So we upload two test firmware to LVFS

<pre>─CHILD MOUSE:
│ │   Device ID:          87835653f7f79d15420fc69988bdadcd623b1314
│ │   Current version:    1.5.6
│ │   Vendor:             USB:0x093A
│ │   GUID:               dbb64836-12f6-519e-a439-904ef0bff587 ← HIDRAW\VID_093A&amp;PID_2801&amp;IDX2
│ │   Device Flags:       • Updatable
│ │                       • Supported on remote server
│ │ 
│ └─PAR2860 Mouse (LVFS) Device Update:
│       New version:      6.5.4
│       Remote ID:        pixart-embargo
│       Summary:          PAR2860 Keyboard for LVFS
│       License:          MIT
│       Size:             43.8 kB
│       Created:          2020-12-12
│       Urgency:          High
│       Vendor:           SDRF-Team-3
│       Flags:            is-upgrade
│       Description:      
│       This stable release fixes the following bugs:
│       
│       • Fix the some hid bug
│       • Fix the some ble bug
│     
├─CHILD KB:
│ │   Device ID:          3b42553c4e3241e8f3f8fbc19a69fa2f95708a9d
│ │   Current version:    1.5.6
│ │   Vendor:             USB:0x093A
│ │   GUID:               569fa7b4-776e-535d-b369-5cab2e4c891d ← HIDRAW\VID_093A&amp;PID_2801&amp;IDX1
│ │   Device Flags:       • Updatable
│ │                       • Supported on remote server
│ │ 
│ └─PAR2860 Mouse (LVFS) Device Update:
│       New version:      6.2.4
│       Remote ID:        pixart-embargo
│       Summary:          PAR2860 Mouse for LVFS
│       License:          MIT
│       Size:             43.8 kB
│       Created:          2021-01-04
│       Urgency:          High
│       Vendor:           SDRF-Team-3
│       Flags:            is-upgrade
│       Description:      
│       This stable release fixes the following bugs:
│       
│       • Fix the some hid bug
│       • Fix the some ble bug
│     
</pre>

I want to know how to update the those child device via his parent ?


<pre>failed to open device for prepare: permission denied
</pre>
 I got the problem  If I update the child device directly (I have set the fd like below)

`fu_udev_device_set_fd (FU_UDEV_DEVICE (dev2), fu_udev_device_get_fd (FU_UDEV_DEVICE (device)));`

or any better way to solve this problem ?

My goal is that we can update fw via USB dongle to wireless mouse or wireless kb. The dongle will get fw bin and bypass  fw to wireless mouse or wireless kb.

**Reply by jyu99** [reply]
@hughsie

**Reply by jyu99** [reply]
It seem like that I create class for wireless module, and rewrite `fu_device_open`.Then get the parent fu-device  in `fu_device_open`  fu_device?

**Reply by hughsie** [reply]
I think the wireless device probably wants to subclass FuDevice (rather than FuUdevDevice) and then when the child->open function also open the parent (`fu_device_get_parent()`) maybe using a FuDeviceLocker -- then issue commands on the parent. e.g. https://github.com/fwupd/fwupd/blob/master/plugins/rts54hid/fu-rts54hid-module.c#L198
